### PR TITLE
Initial support for cloudformation output

### PR DESCRIPTION
**Pull request by justinsb**
Remaining challenges:

* Unclear if we want all resources to be CF managed - particularly resources that we expect to change dynamically, like ASGs.  It might be better only to manage the network in CF.
* We still perform some logic in kops (keypairs, secrets).  The intention was that we want the nodes to be more locked down, though we haven't implemented e.g. KMS which will enforce this distinction - yet.  We need to incorporate this logic into CF.
* We need to move image selection into CF; realistically this probably means selecting a particular image.  However, if we do that it isn't clear how we can update the image e.g. for security updates.

A management node is looking more and more like a safer option.  Most of the complex logic is still in CF - particularly in the networking.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1877)

### CODE REVIEW

**Comment by kris-nova** [comment]
Woooo!

**Comment by kris-nova** [comment]
@justinsb is this still a WIP?

**Comment by justinsb** [comment]
@kris-nova I don't believe so ... it is the equivalent of the terraform output.  I don't think we should actively encourage people to use it until we've figure out the long-term ownership, but I think we could merge it.

We could have a warning I guess, but I prefer to do that in the docs

**Comment by jbeda** [comment]
I want to find time to dig in and really understand this and perhaps sign
up to help move it forward. But it'll be a bit before I have bandwidth.

File: upup/pkg/fi/cloudup/cloudformation/target.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	//	data["provider"] = providersByName
+	//}
+
+	jsonBytes, err := json.MarshalIndent(data, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshalling cloudformation data to json: %v", err)
+	}
+
+	files := make(map[string][]byte)
+	files["kubernetes.json"] = jsonBytes
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/cloudformation/target.go]: I think we are going to want to have a `.yaml` option soon.. just a heads up

Comment (justinsb) [inline upup/pkg/fi/cloudup/cloudformation/target.go]: Well _technically_...

But yes, good point.  Not sure if we should do that as a flag or not.  We could also default to YAML - it feels more readable / usable...

**Comment by kris-nova** [comment]
Also we are going to need to code in delete for this bad boy.. right now you can't delete the stack after it's been created using `kops delete cluster`

FWIW I just deployed a cluster with CF and it works **fabulously**

**Comment by kris-nova** [comment]
@jbeda would it be helpful if we offered a guide like [this one](https://github.com/kubernetes/kops/blob/master/docs/aws.md) that walked a user through creating a cluster with CF + kops?

**Comment by kris-nova** [comment]
@justinsb mind if I PR delete really quick? I think I have it working

File: upup/pkg/fi/cloudup/cloudformation/target.go (hunk 1 above)
Comment (stealthybox) [inline upup/pkg/fi/cloudup/cloudformation/target.go]: ^ I agree with defaulting to yaml
That might be a different change though?

**Comment by kris-nova** [comment]
Also should we have the cloudformation automatically be applied to the cloud? I feel like the  

```
aws cloudformation create-stack --capabilities CAPABILITY_NAMED_IAM --stack-name kubernetes-demo-nivenly-com --template-body file://out/cloudformation/kubernetes.json
```

Might be an extra (and unneeded or unwanted) step?

**Pull request merged**